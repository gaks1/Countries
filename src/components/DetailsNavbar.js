import BackButton from './BackButton';
import styles from './DetailsNavbar.module.css';

const DetailsNavbar = () => (
  <nav className={styles.navbar}>
    <div className={styles.divbutton}>
      <BackButton />
    </div>
    <div>
      <h1 className={styles.h1title}>DETAILS</h1>
    </div>
  </nav>
);

export default DetailsNavbar;
