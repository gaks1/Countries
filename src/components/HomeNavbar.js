import { NavLink } from 'react-router-dom';
import { useState } from 'react';
import PropTypes from 'prop-types';
import styles from './HomeNavbar.module.css';

const Navbar = ({ setSearch }) => {
  const [input, showInput] = useState(false);

  const handleChange = () => {
    showInput(!input);
  };

  return (
    <nav className={styles.navbar}>
      <div>
        <NavLink className={styles.links} to="/">Countries</NavLink>
      </div>
      <div className={styles.header}>
        <div className={styles.headerbutton}>
          <button type="button" onClick={handleChange}>
            <span className="material-symbols-outlined">
              {input ? 'close' : 'search'}
            </span>
          </button>
          {input && (
            <input
              type="text"
              placeholder="Search"
              onChange={(e) => setSearch(e.target.value)}
            />
          )}
        </div>
      </div>
    </nav>
  );
};

Navbar.propTypes = {
  setSearch: PropTypes.func.isRequired,
};

export default Navbar;
