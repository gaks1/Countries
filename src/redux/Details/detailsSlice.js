import {  createSlice } from '@reduxjs/toolkit';

const initialState = {
    countrydetails: [],
    iLoading: false,
    error: null,
};

const detailsSlice = createSlice({
    name: 'details',
    initialState,
    reducers: {},
    extraReducers: {
    },
});

export default detailsSlice.reducer;
