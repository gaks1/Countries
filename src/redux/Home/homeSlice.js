import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';

const url = 'https://restcountries.com/v3.1/all';

const initialState = {
  countries: [],
  isLoading: false,
  fetched: false,
  error: null,
};

export const fetchCountries = createAsyncThunk('home/fetchCountries', async (_, thunkAPI) => {
  try {
    const response = await fetch(`${url}`);
    const data = await response.json();
    return data;
  } catch (error) {
    return thunkAPI.rejectWithValue(error.response.data);
  }
});

const homeSlice = createSlice({
  name: 'home',
  initialState,
  reducers: {},
  extraReducers: {
    [fetchCountries.pending]: (state) => ({ ...state, isLoading: true }),
    [fetchCountries.fulfilled]: (state, action) => ({
      ...state,
      isLoading: false,
      fetched: true,
      countries: action.payload,
    }),
    [fetchCountries.rejected]: (state, action) => ({
      ...state,
      isLoading: false,
      error: action.payload,
    }),
  },

});

export default homeSlice.reducer;
