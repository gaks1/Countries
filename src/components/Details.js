import { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useParams } from 'react-router-dom';
import { fetchCountry } from '../redux/Details/detailsSlice';
import DetailsNavbar from './DetailsNavbar';

const Details = () => {
  const { countryName } = useParams();
  const { countrydetails, isLoading } = useSelector((state) => state.details);
  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(fetchCountry(countryName));
  }, [dispatch, countryName]);

  if (isLoading) {
    return (
      <p>Loading...</p>
    );
  }
  return (
    <div className="Details">
      <h1>dvdv</h1>
      <DetailsNavbar />
      {countrydetails.map(
        (country) => (
          <div key={country.cca3}>
            <div>
              <div key={country.cca3}>
                <div>
                  <img src={country.flags.png} alt={country.flags.alt} />
                </div>
                <div>
                  <h1>{country.name.common}</h1>
                  <p>
                    LAT:
                    {' '}
                    {country.latlng[0]}
                  </p>
                  <p>
                    LONG:
                    {' '}
                    {country.latlng[1]}
                  </p>
                </div>
              </div>
            </div>
            <div>
              <p>COUNTRY DETAILS</p>
            </div>
            <ul>
              <li key={country.cca3}>
                <p>
                  <span>Common name</span>
                  <span>{country.name.common}</span>
                </p>
                <p>
                  <span>Official name</span>
                  <span>{country.name.official}</span>
                </p>
                <p>
                  <span>Capital</span>
                  <span>
                    {Array.isArray(country.capital)
                      ? country.capital.join(',  ')
                      : country.capital}
                  </span>
                </p>
                <p>
                  <span>Country region</span>
                  <span>{country.region}</span>
                </p>
                <p>
                  <span>Country-subregion</span>
                  <span>{country.subregion}</span>
                </p>
                <p>
                  <span>Country-population</span>
                  <span>{country.population}</span>
                </p>
                <p>
                  <span>Country-area</span>
                  <span>{country.area}</span>
                </p>
                <p>
                  <span>Continent</span>
                  <span>{country.continents}</span>
                </p>
                <p>
                  <span>{country.timezones.join(',  ')}</span>
                </p>
                <p>
                  <span>Start of the week</span>
                  <span>{country.startOfWeek}</span>
                </p>
                <p>
                  <span>Street view</span>
                  <a href={country.maps.googleMaps}>google maps</a>
                </p>
                <p>
                  <span>Language:</span>
                  {
                                        country.languages && (
                                        <ul>
                                          {Object.entries(country.languages).map(([code, name]) => (
                                            <li key={code}>{name}</li>
                                          ))}
                                        </ul>
                                        )
                                    }
                </p>

                <p>
                  <span>Currency:</span>
                  {
                                        country.currencies && (
                                        <ul>
                                          {Object.entries(country.currencies)
                                            .map(([code, currency]) => (
                                              <li key={code}>
                                                {currency.name}
                                                {' '}
                                                (
                                                {currency.symbol}
                                                )
                                              </li>
                                            ))}
                                        </ul>
                                        )
                                    }
                </p>

              </li>
            </ul>
          </div>
        ),
      )}

    </div>
  );
};

export default Details;
