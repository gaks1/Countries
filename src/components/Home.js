import { useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { Link } from 'react-router-dom';
import { fetchCountries } from '../redux/Home/homeSlice';

const Home = () => {
  const {
    fetched, isLoading, countries,
  } = useSelector((state) => state.home);
  const dispatch = useDispatch();

  useEffect(() => {
    if (!fetched) {
      dispatch(fetchCountries());
    }
  }, [dispatch, fetched]);

  if (isLoading) {
    return (
      <p>Loading...</p>
    );
  }

  return (
    <div className="home">
      <div>
        <p>LIST OF COUNTRIES</p>
      </div>
      <ul>
        {countries.map((country) => (
          <Link key={country.name.common} to={`/Details/${country.name.common}`}>
            <li key={country.name.common}>
              <span className="material-symbols-outlined">
                arrow_circle_right
              </span>
              <div>
                <img src={country.flags.png} alt={country.flags.alt} />
              </div>
              <div>
                <p>
                  {country.name.common}
                </p>
                <p>
                  LAT:
                  {' '}
                  {country.latlng[0]}
                </p>
                <p>
                  LONG:
                  {' '}
                  {country.latlng[1]}
                </p>
              </div>
            </li>

          </Link>
        ))}
      </ul>
    </div>
  );
};

export default Home;
