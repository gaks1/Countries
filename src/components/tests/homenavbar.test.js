import React from 'react';
import { render } from '@testing-library/react';
import '@testing-library/jest-dom';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import {
  BrowserRouter as Router, Routes, Route, MemoryRouter,
} from 'react-router-dom';

import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import HomeNavbar from '../HomeNavbar';
import DetailsNavbar from '../DetailsNavbar';
import Home from '../Home';
import Details from '../Details';
import filterCountries from '../FilterCountries';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

describe('HomeNavbar', () => {
  const setSearchMock = jest.fn();

  test('renders without crashing', () => {
    const { getByText } = render(
      <Router>
        <HomeNavbar setSearch={setSearchMock} />
      </Router>,
    );
    expect(getByText('Countries')).toBeInTheDocument();
  });

  test('matches snapshot', () => {
    const tree = renderer
      .create(
        <Router>
          <HomeNavbar setSearch={setSearchMock} />
        </Router>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('DetailsNavbar',
  () => {
    test('renders without crashing', () => {
      const { getByText } = render(
        <Router>
          <DetailsNavbar />
        </Router>,
      );
      expect(getByText('DETAILS')).toBeInTheDocument();
    });

    test('matches snapshot', () => {
      const tree = renderer
        .create(
          <Router>
            <DetailsNavbar />
          </Router>,
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

describe('Home', () => {
  const initialState = {
    home: {
      countries: [
        {
          name: {
            common: 'Afghanistan',
          },
          flags: {
            png: 'https://restcountries.com/data/afg.svg',
            alt: 'flag',
          },
          latlng: [20, 20],
        },
        {
          name: {
            common: 'Albania',
          },
          flags: {
            png: 'https://restcountries.com/data/alb.svg',
            alt: 'flag',
          },
          latlng: [20, 20],
        },
      ],
      fetched: false,
      isLoading: false,
    },
  };

  const store = mockStore(initialState);

  test('renders without crashing', () => {
    const { getByText } = render(
      <Provider store={store}>
        <Router>
          <Home />
        </Router>
      </Provider>,
    );
    expect(getByText('Albania')).toBeInTheDocument();
  });

  test('matches snapshot', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <Router>
            <Home />
          </Router>
        </Provider>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('Details', () => {
  const initialState = {
    details: {
      countrydetails: [
        {
          cca3: 'AFG',
          name: {
            common: 'Afghanistan',
            official: 'Islamic Republic of Afghanistan',
          },
          flags: {
            png: 'https://restcountries.com/data/afg.svg',
            alt: 'flag',
          },
          latlng: [20, 20],
          capital: ['Kabul'],
          region: 'Asia',
          subregion: 'Southern Asia',
          population: 32225560,
          area: 652230,
          timezones: ['UTC+04:30'],
          startOfWeek: 'Saturday',
          maps: {
            googleMaps: 'https://www.google.com/maps/place/Afghanistan',
          },
          languages: { ps: 'Pashto', uz: 'Uzbek', tk: 'Turkmen' },
          currencies: {
            AFN: {
              name: 'Afghan afghani',
              symbol: 'Ø‹',
            },
          },
        },
      ],
      isLoading: false,
    },
  };

  const store = mockStore(initialState);
  const countryName = 'Afghanistan';

  test('renders without crashing', () => {
    const { getByText } = render(
      <Provider store={store}>
        <MemoryRouter initialEntries={[`/${countryName}`]}>
          <Routes>
            <Route path="/:countryName" element={<Details />} />
          </Routes>
        </MemoryRouter>
      </Provider>,
    );
    expect(getByText('Saturday')).toBeInTheDocument();
  });

  test('matches snapshot', () => {
    const tree = renderer.create(
      <Provider store={store}>
        <MemoryRouter initialEntries={[`/${countryName}`]}>
          <Routes>
            <Route path="/:countryName" element={<Details />} />
          </Routes>
        </MemoryRouter>
      </Provider>,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('FilterCountries', () => {
  const countries = [
    {
      name: {
        common: 'Afghanistan',
      },
      flags: {
        png: 'https://restcountries.com/data/afg.svg',
        alt: 'flag',
      },
      latlng: [20, 20],
    },
    {
      name: {
        common: 'Albania',
      },
      flags: {
        png: 'https://restcountries.com/data/alb.svg',
        alt: 'flag',
      },
      latlng: [20, 20],
    },
  ];

  test('returns the correct number of countries', () => {
    const filteredCountries = filterCountries({ countries, search: 'Afghanistan' });
    expect(filteredCountries.length).toBe(1);
  });

  test('returns the correct countries', () => {
    const filteredCountries = filterCountries({ countries, search: 'Afghanistan' });
    expect(filteredCountries[0].name.common).toBe('Afghanistan');
  });
});
