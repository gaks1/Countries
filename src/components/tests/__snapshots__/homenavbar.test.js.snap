// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Details matches snapshot 1`] = `
<div
  className="Details"
>
  <nav
    className="navbar"
  >
    <div
      className="divbutton"
    >
      <button
        className="button"
        onClick={[Function]}
        type="button"
      >
        <span
          className="material-symbols-outlined"
        >
          arrow_back_ios
        </span>
      </button>
    </div>
    <div>
      <h1
        className="h1title"
      >
        DETAILS
      </h1>
    </div>
  </nav>
  <div>
    <div
      className="divheader"
    >
      <div
        className="divtitle"
      >
        <div
          className="divtitleimg"
        >
          <img
            alt="flag"
            src="https://restcountries.com/data/afg.svg"
          />
        </div>
        <div
          className="divtitletext"
        >
          <h1>
            Afghanistan
          </h1>
          <p>
            LAT:
             
            20
          </p>
          <p>
            LONG:
             
            20
          </p>
        </div>
      </div>
    </div>
    <div
      className="divseparator"
    >
      <p
        className="septext"
      >
        COUNTRY DETAILS
      </p>
    </div>
    <ul
      className="ul"
    >
      <li
        className="liDetails"
      >
        <p
          className="pdetails"
        >
          <span
            className="pstrong"
          >
            Common name
          </span>
          <span
            className="pproperty"
          >
            Afghanistan
          </span>
        </p>
        <p
          className="pdetails"
        >
          <span
            className="pstrong"
          >
            Official name
          </span>
          <span
            className="pproperty"
          >
            Islamic Republic of Afghanistan
          </span>
        </p>
        <p
          className="pdetails"
        >
          <span
            className="pstrong"
          >
            Capital
          </span>
          <span
            className="pproperty"
          >
            Kabul
          </span>
        </p>
        <p
          className="pdetails"
        >
          <span
            className="pstrong"
          >
            Country region
          </span>
          <span
            className="pproperty"
          >
            Asia
          </span>
        </p>
        <p
          className="pdetails"
        >
          <span
            className="pstrong"
          >
            Country-subregion
          </span>
          <span
            className="pproperty"
          >
            Southern Asia
          </span>
        </p>
        <p
          className="pdetails"
        >
          <span
            className="pstrong"
          >
            Country-population
          </span>
          <span
            className="pproperty"
          >
            32225560
          </span>
        </p>
        <p
          className="pdetails"
        >
          <span
            className="pstrong"
          >
            Country-area
          </span>
          <span
            className="pproperty"
          >
            652230
          </span>
        </p>
        <p
          className="pdetails"
        >
          <span
            className="pstrong"
          >
            Continent
          </span>
          <span
            className="pproperty"
          />
        </p>
        <p
          className="pdetails"
        >
          <span
            className="pstrong"
          >
            Timezones
          </span>
          <span
            className="pproperty"
          >
            UTC+04:30
          </span>
        </p>
        <p
          className="pdetails"
        >
          <span
            className="pstrong"
          >
            Start of the week
          </span>
          <span
            className="pproperty"
          >
            Saturday
          </span>
        </p>
        <p
          className="pdetails"
        >
          <span
            className="pstrong"
          >
            Street view
          </span>
          <a
            className="pproperty"
            href="https://www.google.com/maps/place/Afghanistan"
          >
            google maps
          </a>
        </p>
        <p
          className="pdetails"
        >
          <span
            className="pstrong"
          >
            Language:
          </span>
          <ul
            className="languageContainer"
          >
            <li
              className="languageItem"
            >
              Pashto
            </li>
            <li
              className="languageItem"
            >
              Uzbek
            </li>
            <li
              className="languageItem"
            >
              Turkmen
            </li>
          </ul>
        </p>
        <p
          className="pdetails"
        >
          <span
            className="pstrong"
          >
            Currency:
          </span>
          <ul
            className="languageContainer"
          >
            <li
              className="languageItem"
            >
              Afghan afghani
               
              (
              Ø‹
              )
            </li>
          </ul>
        </p>
      </li>
    </ul>
  </div>
</div>
`;

exports[`DetailsNavbar matches snapshot 1`] = `
<nav
  className="navbar"
>
  <div
    className="divbutton"
  >
    <button
      className="button"
      onClick={[Function]}
      type="button"
    >
      <span
        className="material-symbols-outlined"
      >
        arrow_back_ios
      </span>
    </button>
  </div>
  <div>
    <h1
      className="h1title"
    >
      DETAILS
    </h1>
  </div>
</nav>
`;

exports[`Home matches snapshot 1`] = `
<div
  className="home"
>
  <nav
    className="navbar"
  >
    <div>
      <a
        aria-current="page"
        className="links active"
        href="/"
        onClick={[Function]}
      >
        Countries
      </a>
    </div>
    <div
      className="header"
    >
      <div
        className="headerbutton"
      >
        <button
          onClick={[Function]}
          type="button"
        >
          <span
            className="material-symbols-outlined"
          >
            search
          </span>
        </button>
      </div>
    </div>
  </nav>
  <div
    className="divheader"
  >
    <div
      className="divtitle"
    >
      <div
        className="imageContainer"
      >
        <img
          alt=""
          src="Global-Map-Transparent-700x416.png"
        />
      </div>
      <div
        className="textContainer"
      >
        <h1>
          COUNTRIES AROUND
          <br />
          THE WORLD
        </h1>
      </div>
    </div>
  </div>
  <div
    className="separator"
  >
    <p
      className="septext"
    >
      LIST OF COUNTRIES
    </p>
  </div>
  <ul
    className="ulCountries"
  >
    <a
      className="link"
      href="/Details/Afghanistan"
      onClick={[Function]}
    >
      <li
        className="liCountry"
      >
        <span
          className="material-symbols-outlined arrowCircleRight"
        >
          arrow_circle_right
        </span>
        <div
          className="countryImage"
        >
          <img
            alt="flag"
            src="https://restcountries.com/data/afg.svg"
          />
        </div>
        <div
          className="countryDetails"
        >
          <p
            className="liname"
          >
            Afghanistan
          </p>
          <p
            className="licood"
          >
            LAT:
             
            20
          </p>
          <p
            className="licood"
          >
            LONG:
             
            20
          </p>
        </div>
      </li>
    </a>
    <a
      className="link"
      href="/Details/Albania"
      onClick={[Function]}
    >
      <li
        className="liCountry"
      >
        <span
          className="material-symbols-outlined arrowCircleRight"
        >
          arrow_circle_right
        </span>
        <div
          className="countryImage"
        >
          <img
            alt="flag"
            src="https://restcountries.com/data/alb.svg"
          />
        </div>
        <div
          className="countryDetails"
        >
          <p
            className="liname"
          >
            Albania
          </p>
          <p
            className="licood"
          >
            LAT:
             
            20
          </p>
          <p
            className="licood"
          >
            LONG:
             
            20
          </p>
        </div>
      </li>
    </a>
  </ul>
</div>
`;

exports[`HomeNavbar matches snapshot 1`] = `
<nav
  className="navbar"
>
  <div>
    <a
      aria-current="page"
      className="links active"
      href="/"
      onClick={[Function]}
    >
      Countries
    </a>
  </div>
  <div
    className="header"
  >
    <div
      className="headerbutton"
    >
      <button
        onClick={[Function]}
        type="button"
      >
        <span
          className="material-symbols-outlined"
        >
          search
        </span>
      </button>
    </div>
  </div>
</nav>
`;
